src/SearchWidget.cpp

```diff
@@ -0,0 +1,181 @@
+
+    QObject::connect(mUi->guiNextButton, &QToolButton::clicked, this, &SearchWidget::onNextButtonClicked);
+    QObject::connect(mUi->guiPreviousButton, &QToolButton::clicked, this, &SearchWidget::onPreviousButtonClicked);
+    QObject::connect(mTextEdit, &TextEdit::textChanged, this, &SearchWidget::documentChange);
+    QObject::connect(mUi->guiSearchLine, &QLineEdit::textChanged, this, &SearchWidget::searchLineChanged);
+    QObject::connect(mUi->guiCloseButton, &QToolButton::clicked, this, &SearchWidget::closeSearch);
+}
+
+SearchWidget::~SearchWidget()
+{
+    delete mUi;
+}
+
+void SearchWidget::initialize(const QString & text)
+{
+    mSearchVisible = true;
+    mUi->guiSearchLine->setFocus();
+    mUi->guiSearchLine->setText(text);
+    searchWord(true);
+}
+
+void SearchWidget::uninitialize()
+{
+    mSearchVisible = false;
+    highLightedWords(false);
+}
+
+
+void SearchWidget::keyPressEvent(QKeyEvent * event)
+{
+    if (event->key() == Qt::Key_Return || event->key() == Qt::Key_Enter) {
+        onNextButtonClicked();
+    }
+}
+
+void SearchWidget::searchWord(bool selectNext, QString searchValue)
+{
+    if (searchValue.isEmpty()) {
+        searchValue = mUi->guiSearchLine->text();
+    }
+    mTextDocument = mTextEdit->toPlainText();
+    mCurrentSelectedWord = -1;
+    mPositionWords.clear();
+    QTextDocument *document = mTextEdit->document();
+    searchPositionsWordsInDocument(searchValue, selectNext, document);
+}
+
+void SearchWidget::onNextButtonClicked()
+{
+    if (mPositionWords.empty()) {
+        return;
+    }
+    if (mCurrentSelectedWord != ((int)mPositionWords.size() - 1)) {
+        mCurrentSelectedWord++;
+    }
+    else {
+        mCurrentSelectedWord = 0;
+    }
+    QTextDocument *document = mTextEdit->document();
+    QTextCursor cursor(document);
+    cursor.beginEditBlock();
+    int wordStartPosition = mPositionWords.at(mCurrentSelectedWord);
+    cursor.setPosition(wordStartPosition, QTextCursor::MoveAnchor);
+    cursor.setPosition(wordStartPosition + mUi->guiSearchLine->text().size(), QTextCursor::KeepAnchor);
+    mTextEdit->setTextCursor(cursor);
+    cursor.endEditBlock();
+    setCountAndCurrentPosition();
+}
+
+void SearchWidget::onPreviousButtonClicked()
+{
+    if (mPositionWords.empty()) {
+        return;
+    }
+    if (mCurrentSelectedWord != 0) {
+        mCurrentSelectedWord--;
+    }
+    else {
+        mCurrentSelectedWord = mPositionWords.size() - 1;
+    }
+    QTextDocument *document = mTextEdit->document();
+    QTextCursor cursor(document);
+    cursor.beginEditBlock();
+    int wordStartPosition = mPositionWords.at(mCurrentSelectedWord);
+    cursor.setPosition(wordStartPosition, QTextCursor::MoveAnchor);
+    cursor.setPosition(wordStartPosition + mUi->guiSearchLine->text().size(), QTextCursor::KeepAnchor);
+    mTextEdit->setTextCursor(cursor);
+    cursor.endEditBlock();
+    setCountAndCurrentPosition();
+}
```

Comment:
There is a lot of duplication between `onNextButtonClicked()` and `onPreviousButtonClicked()`.
You should be able to factorize all the code after the update of mCurrentSelectedWord between the two methods.

Reply:
resolve in commit baad493